nemo/core/neural_graph.py

```diff
@@ -962,3 +967,79 @@ def unfreeze(self, module_names: Optional[List[str]] = None):
+            Args:
+                filename (string): Name of the file where the checkpoint will be saved.
+                module_names: List of modules to be frozen (Optional). If passed, all modules will be saved.
+            Raises:
+                KeyError: If name of the module won't be recognized.
+        """
+        # Work on all modules.
+        if module_names is None:
+            module_names = self._modules.keys()
+
+        # Prepare the "graph checkpoint".
+        chkpt = {"header": {}, "modules": {}}
+        chkpt["header"] = {"nemo_core_version": nemo_version, "name": self.name}
+
+        log_str = ''
+        # Iterate through the modules one by one.
+        for name in module_names:
+            if name not in self._modules.keys():
+                raise KeyError("Module `{}` not present in the `{}` graph".format(name, self.name))
+            # Check module type.
+            module = self._modules[name]
+            if module.type == ModuleType.trainable:
+                # Get module state_dict().
+                chkpt["modules"][name] = get_state_dict(module)
+                log_str += "  * Module '{}' ({}) params saved \n".format(module.name, type(module).__name__)
+            else:
+                logging.debug("Module `{}` is not trainable so cannot be saved".format(name))
+
+        # Save checkpoint.
+        save(chkpt, filename)
+        log_str = "Saved  the '{}' graph to a checkpoint `{}`:\n".format(self.name, filename) + log_str
+        logging.info(log_str)
+
+    def restore_from(self, filename: str, module_names: Optional[List[str]] = None):
+        """
+            Restores the state of trainable modules in the graph from a checkpoint file.
+            Args:
+                filename (string): Name of the checkpoint to be restored from.
+                module_names: List of modules to be frozen (Optional). If passed, all modules will be restored.
+            Raises:
+                KeyError: If name of the module won't be recognized.
+        """
+        # Work on all modules.
+        if module_names is None:
+            module_names = self._modules.keys()
+
+        # Load the checkpoint.
+        chkpt = load(filename)
+
+        log_str = "Loading modules constituting the '{}' graph from the `{}` checkpoint :\n".format(
+            chkpt["header"]["name"], filename
+        )
+
+        # Iterate through the modules one by one.
+        for name in module_names:
+            try:
+                # Get module.
+                module = self._modules[name]
+                # Restore module weights
+                set_state_dict(module, chkpt["modules"][name])
+                log_str += "  * Module '{}' ({}) params loaded\n".format(module.name, type(module).__name__)
+            except KeyError:
+                log_str += "  ! Module '{}' params not found in checkpoint\n".format(name)
+                warning = True
```

Comment:
Default value of `warning` needs to be declared outside the loop, in the edge case that warning is never defined and the loop completes and the below `if warning:` check executes.

Reply:
of course! good catch! :)